sdks/python/apache_beam/transforms/util.py

```diff
@@ -815,18 +817,25 @@ def __init__(self, batch_size, max_buffering_duration_secs=None):
     _shard_id_prefix = uuid.uuid4().bytes
 
     def expand(self, pcoll):
+      key_type, value_type = pcoll.element_type.tuple_types
       sharded_pcoll = pcoll | Map(
           lambda key_value: (
               ShardedKey(
                   key_value[0],
                   # Use [uuid, thread id] as the shard id.
                   GroupIntoBatches.WithShardedKey._shard_id_prefix + bytes(
                       threading.get_ident().to_bytes(8, 'big'))),
-              key_value[1]))
+              key_value[1])).with_output_types(
+                  TupleConstraint([ShardedKeyType[key_type], value_type]))
       return (
           sharded_pcoll
           | GroupIntoBatches(self.batch_size, self.max_buffering_duration_secs))
 
+    def infer_output_type(self, input_type):
```

Comment:
If you keep the with_output_types decorator, please remove this.

Reply:
Couldn't use the decorator due to the error I mentioned in another comment.